nanobenchmarks/main.cpp

```diff
@@ -0,0 +1,268 @@
+
+#include <algorithm>
+#include <iostream>
+#include <iomanip>
+#include <limits>
+#include <map>
+#include <memory>
+#include <print>
+#include <ranges>
+#include <regex>
+#include <sstream>
+#include <string>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+#include <intrin.h>
+
+#include "absl/flags/flag.h"
+#include "absl/flags/parse.h"
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_split.h"
+#include "base/array.hpp"
+#include "base/cpuid.hpp"
+#include "mathematica/logger.hpp"
+#include "nanobenchmarks/function_registry.hpp"
+#include "nanobenchmarks/microarchitectures.hpp"
+#include "nanobenchmarks/performance_settings_controller.hpp"
+#include "numerics/cbrt.hpp"
+#include "testing_utilities/statistics.hpp"
+
+ABSL_FLAG(std::size_t,
+          loop_iterations,
+          100,
+          "Number of iterations of the measured loop");
+ABSL_FLAG(std::size_t,
+          samples,
+          1'000'000,
+          "Number of measurements to perform for each benchmarked function");
+ABSL_FLAG(std::string,
+          benchmark_filter,
+          ".*",
+          "Regular expression matching the names of functions to benchmark");
+ABSL_FLAG(std::string,
+          log_to_mathematica,
+          "",
+          "File to which to log the measurements");
+ABSL_FLAG(double, input, 5, "Input for the benchmarked functions");
+
+// Adding support for flag types only works using ADL (or by being in
+// marshalling.h), so we do this, which is UB.
+namespace std {
+
+bool AbslParseFlag(absl::string_view const text,
+                   std::vector<double>* const flag,
+                   std::string* const error) {
+  flag->clear();
+  for (absl::string_view const element : absl::StrSplit(text, ',')) {
+    if (!absl::ParseFlag(element, &flag->emplace_back(), error)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+std::string AbslUnparseFlag(std::vector<double> const& flag) {
+  return absl::StrJoin(flag, ",");
+}
+
+}  // namespace std
+
+ABSL_FLAG(std::vector<double>,
+          quantiles,
+          (std::vector<double>{0.001, 0.01, 0.05, 0.1, 0.25, 0.5}),
+          "Quantiles to report");
+
+namespace principia {
+namespace nanobenchmarks {
+namespace _main {
+namespace {
+
+using namespace principia::mathematica::_logger;
+using namespace principia::mathematica::_mathematica;
+using namespace principia::nanobenchmarks::_function_registry;
+using namespace principia::nanobenchmarks::_microarchitectures;
+using namespace principia::nanobenchmarks::_performance_settings_controller;
+using namespace principia::testing_utilities::_statistics;
+
+struct LatencyDistributionTable {
```

Comment:
Let's move this struct to its own file with proper `hpp`/`cpp` separation.  Also, maybe make it a `class`?

Reply:
The logic is all formatting, which is coupled to main, and the structure is tied to the Benchmark function. Let’s untangle that some other day.